httpcore/_backends/sync.py

```diff
@@ -17,6 +19,117 @@
 from .base import SOCKET_OPTION, NetworkBackend, NetworkStream
 
 
+class SyncTLSStream(NetworkStream):
+    """
+    Because the standard `SSLContext.wrap_socket` method does
+    not work for `SSLSocket` objects, we need this class
+    to implement TLS stream using an underlying `SSLObject`
+    instance in order to support TLS on top of TLS.
+    """
+
+    # Defined in RFC 8449
+    TLS_RECORD_SIZE = 16384
+
+    def __init__(
+        self,
+        sock: socket.socket,
+        ssl_context: ssl.SSLContext,
+        server_hostname: typing.Optional[str] = None,
+        timeout: typing.Optional[float] = None,
+    ):
+        self._sock = sock
+        self._incoming = ssl.MemoryBIO()
+        self._outgoing = ssl.MemoryBIO()
+
+        self.ssl_obj = ssl_context.wrap_bio(
+            incoming=self._incoming,
+            outgoing=self._outgoing,
+            server_hostname=server_hostname,
+        )
+
+        self._perform_io(self.ssl_obj.do_handshake, timeout)
+
+    def _perform_io(
+        self,
+        func: typing.Callable[..., typing.Any],
+        timeout: typing.Optional[float],
+    ) -> typing.Any:
+        ret = None
+        timeout = timeout or None  # Replaces `0` with `None`
+
+        while True:
+            errno = None
+            try:
+                ret = func()
+            except (ssl.SSLWantReadError, ssl.SSLWantWriteError) as e:
+                errno = e.errno
+
+            if timeout is not None and timeout <= 0:  # pragma: no cover
+                raise socket.timeout()
+
+            self._sock.settimeout(timeout)
+            operation_start = perf_counter()
+            self._sock.sendall(self._outgoing.read())
+            # If the timeout is `None`, don't touch it.
+            timeout = timeout and timeout - (perf_counter() - operation_start)
```

Comment:
```suggestion
            timeout = None if timeout is None else timeout - (perf_counter() - operation_start)
```

here during next rounds, it can be exactly equals to zero, so is better to use condition here.

Reply:
Oh correct, actually that that was also my suggestion to raise for `timeout <= 0` before calls 😅